From 9d1983114d4cbc28359f3802a953e6942132f63f Mon Sep 17 00:00:00 2001
From: Edgar E. Iglesias <edgar.iglesias@gmail.com>
Date: Thu, 17 Nov 2011 23:44:38 +0100
Subject: [PATCH 11/24] microblaze: Correct adjustment of global symbols after linker relaxation

Signed-off-by: Edgar E. Iglesias <edgar.iglesias@gmail.com>
---
 bfd/elf32-microblaze.c |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/bfd/elf32-microblaze.c b/bfd/elf32-microblaze.c
index b8c65b5..23c88a3 100644
--- a/bfd/elf32-microblaze.c
+++ b/bfd/elf32-microblaze.c
@@ -1783,12 +1783,12 @@ microblaze_elf_relax_section (bfd *abfd,
 
       /* Now adjust the global symbols defined in this section.  */
       isym = isymbuf + symtab_hdr->sh_info;
+      symcount =  (symtab_hdr->sh_size / sizeof (Elf32_External_Sym)) - symtab_hdr->sh_info;
       isymend = isymbuf + (symtab_hdr->sh_size / sizeof (Elf32_External_Sym));
-      for (sym_index = 0; isym < isymend; isym++, sym_index++)
+      for (sym_index = 0; sym_index < symcount; sym_index++)
         {
           sym_hash = elf_sym_hashes (abfd)[sym_index];
-          if (isym->st_shndx == shndx
-              && (sym_hash->root.type == bfd_link_hash_defined
+          if ((sym_hash->root.type == bfd_link_hash_defined
                   || sym_hash->root.type == bfd_link_hash_defweak)
               && sym_hash->root.u.def.section == sec)
 	    {
-- 
1.7.1

