From d764ea7a192c1e86ea95bb75e6fb73b30f43e413 Mon Sep 17 00:00:00 2001
From: Edgar E. Iglesias <edgar.iglesias@gmail.com>
Date: Wed, 25 Apr 2012 15:08:28 +0200
Subject: [PATCH 23/41] microblaze: Correct sigcontextinfo

Signed-off-by: Edgar E. Iglesias <edgar.iglesias@gmail.com>
---
 .../unix/sysv/linux/microblaze/sigcontextinfo.h    |    8 ++++----
 1 files changed, 4 insertions(+), 4 deletions(-)

diff --git a/sysdeps/unix/sysv/linux/microblaze/sigcontextinfo.h b/sysdeps/unix/sysv/linux/microblaze/sigcontextinfo.h
index 6a4399c..12dbadc 100644
--- a/sysdeps/unix/sysv/linux/microblaze/sigcontextinfo.h
+++ b/sysdeps/unix/sysv/linux/microblaze/sigcontextinfo.h
@@ -13,10 +13,10 @@
    Software Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA
    02111-1307 USA.  */
 
-#define SIGCONTEXT int _code, struct sigcontext *
+#define SIGCONTEXT int _code, struct ucontext *
 #define SIGCONTEXT_EXTRA_ARGS _code,
-#define GET_PC(ctx)	((void *) (ctx)->regs.pc)
-#define GET_FRAME(ctx)	((void *) (ctx)->regs.sp)
-#define GET_STACK(ctx)	((void *) (ctx)->regs.sp)
+#define GET_PC(ctx)    ((void *) (ctx)->uc_mcontext.regs.pc)
+#define GET_FRAME(ctx) ((void *) (ctx)->uc_mcontext.regs.sp)
+#define GET_STACK(ctx) ((void *) (ctx)->uc_mcontext.regs.sp)
 #define CALL_SIGHANDLER(handler, signo, ctx) \
   (handler)((signo), SIGCONTEXT_EXTRA_ARGS (ctx))
-- 
1.7.1

