From 46f16f36a406646b685240e2f3925feca5ed7b2e Mon Sep 17 00:00:00 2001
From: David Holsgrove <david.holsgrove@petalogix.com>
Date: Thu, 10 May 2012 14:41:55 +1000
Subject: [PATCH 31/41] microblaze: Add target to configure.in template

Signed-off-by: David Holsgrove <david.holsgrove@petalogix.com>
---
 configure.in |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/configure.in b/configure.in
index ff5a121..37174d1 100644
--- a/configure.in
+++ b/configure.in
@@ -544,6 +544,7 @@ i[34567]86)	base_machine=i386 machine=i386/$machine ;;
 ia64)		base_machine=ia64 machine=ia64 ;;
 m88???)		base_machine=m88k machine=m88k/$machine ;;
 m88k)		base_machine=m88k machine=m88k/m88100 ;;
+microblaze*)    base_machine=microblaze machine=microblaze ;;
 powerpc)	base_machine=powerpc machine=powerpc/powerpc32 ;;
 powerpc64)	base_machine=powerpc machine=powerpc/powerpc64 ;;
 s390)           base_machine=s390 machine=s390/s390-32 ;;
-- 
1.7.1

