From 30542dbc9cc31dcc4281dd1222ea3702f99af0de Mon Sep 17 00:00:00 2001
From: Edgar E. Iglesias <edgar.iglesias@gmail.com>
Date: Tue, 18 Oct 2011 22:40:35 +0200
Subject: [PATCH 11/79] microblaze: Add little-endian multilib option

Signed-off-by: Edgar E. Iglesias <edgar.iglesias@gmail.com>
---
 gcc/config/microblaze/t-microblaze |    6 ++++--
 1 files changed, 4 insertions(+), 2 deletions(-)

diff --git a/gcc/config/microblaze/t-microblaze b/gcc/config/microblaze/t-microblaze
index d926b1f..99f31c7 100644
--- a/gcc/config/microblaze/t-microblaze
+++ b/gcc/config/microblaze/t-microblaze
@@ -3,9 +3,11 @@ EXTRA_MULTILIB_PARTS = crtbegin$(objext) crtend$(objext)
 
 EXTRA_PARTS += crti$(objext) crtn$(objext)
 
-MULTILIB_OPTIONS = mxl-barrel-shift mno-xl-soft-mul mxl-multiply-high
-MULTILIB_DIRNAMES = bs m mh
+MULTILIB_OPTIONS = mxl-barrel-shift mno-xl-soft-mul mxl-multiply-high mlittle-endian
+MULTILIB_DIRNAMES = bs m mh le
 MULTILIB_EXCEPTIONS = *mxl-barrel-shift/mxl-multiply-high mxl-multiply-high
+MULTILIB_EXCEPTIONS += *mxl-barrel-shift/mxl-multiply-high/mlittle-endian
+MULTILIB_EXCEPTIONS += mxl-multiply-high/mlittle-endian
 
 # Extra files
 microblaze-c.o: $(srcdir)/config/microblaze/microblaze-c.c \
-- 
1.7.1

