From 7592ce809e77ee2273b0f71d13d7b9c6782ed6bb Mon Sep 17 00:00:00 2001
From: David Holsgrove <david.holsgrove@petalogix.com>
Date: Tue, 15 May 2012 19:47:30 +1000
Subject: [PATCH 58/79] microblaze: port patch from crosstool-ng/patches/gcc/4.6.2/100-libgcc_eh.a.patch

Signed-off-by: David Holsgrove <david.holsgrove@petalogix.com>
---
 libgcc/Makefile.in |   11 ++++++-----
 1 files changed, 6 insertions(+), 5 deletions(-)

diff --git a/libgcc/Makefile.in b/libgcc/Makefile.in
index b57aeb6..59db267 100644
--- a/libgcc/Makefile.in
+++ b/libgcc/Makefile.in
@@ -772,8 +772,9 @@ all: libunwind.a
 libgcc_s$(SHLIB_EXT): libunwind$(SHLIB_EXT)
 endif
 
+all: libgcc_eh.a
 ifeq ($(enable_shared),yes)
-all: libgcc_eh.a libgcc_s$(SHLIB_EXT)
+all: libgcc_s$(SHLIB_EXT)
 ifneq ($(LIBUNWIND),)
 all: libunwind$(SHLIB_EXT)
 endif
@@ -950,10 +951,6 @@ install-libunwind:
 install-shared:
 	$(mkinstalldirs) $(DESTDIR)$(inst_libdir)
 
-	$(INSTALL_DATA) libgcc_eh.a $(DESTDIR)$(inst_libdir)/
-	chmod 644 $(DESTDIR)$(inst_libdir)/libgcc_eh.a
-	$(RANLIB) $(DESTDIR)$(inst_libdir)/libgcc_eh.a
-
 	$(subst @multilib_dir@,$(MULTIDIR),$(subst \
 		@shlib_base_name@,libgcc_s,$(subst \
 		@shlib_slibdir_qual@,$(MULTIOSSUBDIR),$(SHLIB_INSTALL))))
@@ -968,6 +965,10 @@ install-leaf: $(install-shared) $(install-libunwind)
 	chmod 644 $(DESTDIR)$(inst_libdir)/libgcov.a
 	$(RANLIB) $(DESTDIR)$(inst_libdir)/libgcov.a
 
+	$(INSTALL_DATA) libgcc_eh.a $(DESTDIR)$(inst_libdir)/
+	chmod 644 $(DESTDIR)$(inst_libdir)/libgcc_eh.a
+	$(RANLIB) $(DESTDIR)$(inst_libdir)/libgcc_eh.a
+
 	parts="$(INSTALL_PARTS)";				\
 	for file in $$parts; do					\
 	  rm -f $(DESTDIR)$(inst_libdir)/$$file;		\
-- 
1.7.1

